### ISSUE TITLE
Add skip-grams to tokenize

### ISSUE THREAD

**Issue by zachmayer** [kbenoit/quanteda#53]
Would work similar to ngrams (e.g. `skips=1` or `skips=2:3`), but would add [skip-grams](https://en.wikipedia.org/wiki/N-gram#Skip-gram), rather than n-grams.

**Comment by kbenoit** [comment]
Working now, see `?skipgrams`. e.g.

    > test <- "this is just a test text i'm using"
    > skipgrams(tokenize(test), concatenator = " ")
    [[1]]
     [1] "this is"    "is just"    "just a"     "a test"     "test text"  "text i'm"   "i'm using"  "this just"  "is a"      
    [10] "just test"  "a text"     "test i'm"   "text using"

**Comment by zachmayer** [comment]
Awesome, thanks!

**Comment by kbenoit** [comment]
Welcome. Will try to start closing more issues as I have some time this month. If you use the skipgrams to get an R version 9or even partial R version) of word2vec working, please let me know, we'd love to add this functionality to quanteda.

**Comment by zachmayer** [comment]
I've got my own implementation of word2vec partially implemented.  This blog post does an excellent job explaining how it works: http://rare-technologies.com/making-sense-of-word2vec/

Basically, you can get a good approximation of word2vec with skip-grams + SVD

**Comment by zachmayer** [comment]
Also, you should be aware of my [r2vec package](https://github.com/zachmayer/r2vec) and the [tmlite package](https://github.com/dselivanov/tmlite).

I'm planning to remove a lot of the tokenization and n-grams stuff from r2vec (provided quanreda and tmlite provide fast skip gram tokenizers and prediction methods for new data), and focus on a good implementation of word2vec and doc2vec.

**Comment by kbenoit** [comment]
I'd be happy to work with you on integrating with r2vec. quanteda is faster anyway: see https://github.com/dselivanov/tmlite/issues/23. And the API is cleaner. r2vec::textVectors() is impressive but - if I may say so - too much functionality for one function. Only the pca functionality cannot be done in quanteda. Why not use quanteda to go from corpus to dfm, with skipgram tokenization, and weighting and trimming  the dfm in a few additional steps, and then define your text methods to work on weighted dfm objects? Simpler for users and simpler to program. I'd be happy to contribute.

**Comment by zachmayer** [comment]
This issue is still the primary reason I'm maintaining my own code: https://github.com/kbenoit/quanteda/issues/46

In general prediction pipelines are a huge issue for me.  E.g. if I have a tokenization -> skipgrams -> word2vec -> pca pipeline built on a training set of data, I need to be able to easily project test set documents into the exact same vector space.

**Comment by kbenoit** [comment]
OK then I will work on that as a priority!

Ken


On 6 Oct 2015, at 16:52, Zach Mayer <notifications@github.com<mailto:notifications@github.com>> wrote:


This issue is still the primary reason I'm maintaining my own code: #46<https://github.com/kbenoit/quanteda/issues/46>

In general prediction pipelines are a huge issue for me. E.g. if I have a tokenization -> skipgrams -> word2vec -> pca pipeline built on a training set of data, I need to be able to easily project test set documents into the exact same vector space.

**Comment by zachmayer** [comment]
Sounds good, thanks!

**Comment by kbenoit** [comment]
If you're happy with the solution to #46 then please close this issue and #46.
